rustv1/examples/cloudwatchlogs/src/bin/large-query.rs

```diff
@@ -0,0 +1,436 @@
+        let response = self
+            .client
+            .start_query()
+            .log_group_name(self.log_group_name.clone())
+            .query_string("fields @timestamp, @message | sort @timestamp asc")
+            .start_time(date_range.0.timestamp())
+            .end_time(date_range.1.timestamp())
+            .limit(self.limit)
+            .send()
+            .await;
+        match response {
+            Ok(start) => Ok(start.query_id.expect("start query query_id")),
+            Err(err) => {
+                if err
+                    .message()
+                    .unwrap_or_default()
+                    .starts_with("Query's end date and time")
+                {
+                    Err(LargeQueryError::DateOutOfBounds)
+                } else {
+                    Err(LargeQueryError::FromCloudwatchLogs(err.into()))
+                }
+            }
+        }
+    }
+}
+
+fn get_last_log_date(results: &[Vec<ResultField>]) -> Result<DateTime<Utc>, LargeQueryError> {
+    let last = results
+        .last()
+        .expect("last query item")
+        .iter()
+        .find(|e| matches!(e.field(), Some("@timestamp")))
+        .expect("timestamp field")
+        .value()
+        .expect("timestamp field value");
+
+    DateTime::parse_from_rfc3339(format!("{last}Z").as_str())
+        .map(|e| e.to_utc())
+        .map_err(LargeQueryError::FromChronoParse)
+}
+
+#[tokio::main]
+async fn main() -> Result<(), LargeQueryError> {
+    tracing_subscriber::fmt::init();
+
+    let group = env::var("QUERY_GROUP").unwrap_or("/workflows/cloudwatch-logs/large-query".into());
+    let start_date = DateTime::from_timestamp_millis(
+        env::var("QUERY_START_DATE")
+            .expect("lookup QUERY_START_DATE")
+            .parse()
+            .unwrap(),
+    )
+    .expect("parse start date")
+    .with_timezone(&Utc);
+    let end_date = DateTime::from_timestamp_millis(
+        env::var("QUERY_END_DATE")
+            .expect("lookup QUERY_END_DATE")
+            .parse()
+            .unwrap(),
+    )
+    .expect("parse end date")
+    .with_timezone(&Utc);
+
+    println!("{group}");
+    println!("{start_date}");
+    println!("{end_date}");
+
+    let shared_config = aws_config::from_env()
+        .behavior_version(BehaviorVersion::latest())
+        .load()
+        .await;
+    let client = Client::new(&shared_config);
+
+    let mut query = CloudWatchLongQuery::new(client, group, DateRange(start_date, end_date));
+
+    query.run().await.expect("run query to completion");
+
+    eprintln!("Total results: {}", query.results.len());
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use aws_config::Region;
+    use aws_sdk_cloudwatchlogs::{operation::start_query::StartQueryOutput, Config};
+    use aws_smithy_mocks_experimental::{mock, MockResponseInterceptor, Rule};
+    use chrono::{TimeZone, Utc};
+
+    fn from_rules(rules: &[&Rule], enforce_order: bool) -> Client {
```

Comment:
might rename this function to clarify that the thing you're making is a client

Reply:
Done